subsys/net/lib/nrf_cloud/src/nrf_cloud.c

```diff
@@ -408,44 +462,149 @@ static int send(const struct cloud_backend *const backend,
 	return 0;
 }
 
-static int ping(const struct cloud_backend *const backend)
+static int api_ping(const struct cloud_backend *const backend)
 {
 	/* TODO: Do only ping, nrf_cloud_process() also checks for input. */
 	nrf_cloud_process();
 
 	return 0;
 }
 
-static int keepalive_time_left(const struct cloud_backend *const backend)
+static int api_keepalive_time_left(const struct cloud_backend *const backend)
 {
 	return nct_keepalive_time_left();
 }
 
-static int input(const struct cloud_backend *const backend)
+static int api_input(const struct cloud_backend *const backend)
 {
 	nrf_cloud_process();
 
 	return 0;
 }
 
-static int user_data_set(const struct cloud_backend *const backend,
+static int api_user_data_set(const struct cloud_backend *const backend,
 			 void *user_data)
 {
 	backend->config->user_data = user_data;
 
 	return 0;
 }
 
+void connection_poll(void)
+{
+	int ret;
+	struct pollfd fds[1];
+	struct cloud_event cloud_evt = {
+		.type = CLOUD_EVT_DISCONNECTED,
+		.data = { .err = CLOUD_DISCONNECT_ERR_MISC}
+	};
+
+start:
+	k_sem_take(&connection_poll_sem, K_FOREVER);
+	atomic_set(&connection_poll_active, 1);
+
+	cloud_evt.data.err = CLOUD_CONNECT_RES_SUCCESS;
+	cloud_evt.type = CLOUD_EVT_CONNECTING;
+	cloud_notify_event(nrf_cloud_backend, &cloud_evt, NULL);
+
+	ret = nrf_cloud_connect(NULL);
+	ret = translate_connect_error(ret);
+
+	if (ret != CLOUD_CONNECT_RES_SUCCESS) {
+		cloud_evt.data.err = ret;
+		cloud_evt.type = CLOUD_EVT_CONNECTING;
+		cloud_notify_event(nrf_cloud_backend, &cloud_evt, NULL);
+		goto reset;
+	} else {
+		LOG_DBG("Cloud connection request sent.");
+	}
+
+	fds[0].fd = nct_socket_get();
+	fds[0].events = POLLIN;
+
+	/* Only disconnect events will occur below */
+	cloud_evt.type = CLOUD_EVT_DISCONNECTED;
+
+	while (true) {
+		ret = poll(fds, ARRAY_SIZE(fds), POLL_TIMEOUT_MS);
+
+		if (atomic_get(&disconnect_requested)) {
+			atomic_set(&disconnect_requested, 0);
+			LOG_DBG("Expected disconnect event.");
+			cloud_evt.data.err = CLOUD_DISCONNECT_ERR_USER_REQUEST;
+			cloud_notify_event(nrf_cloud_backend, &cloud_evt, NULL);
+			goto reset;
+		}
+
+		if (ret < 0) {
+			LOG_DBG("poll() returned an error: %d", ret);
+			cloud_evt.data.err = CLOUD_DISCONNECT_ERR_MISC;
+			break;
+		}
+
+		if (ret == 0) {
+			if (cloud_keepalive_time_left(nrf_cloud_backend) <
+			    POLL_TIMEOUT_MS) {
+				api_ping(nrf_cloud_backend);
+			}
+			continue;
+		}
+
+		if ((fds[0].revents & POLLIN) == POLLIN) {
+			api_input(nrf_cloud_backend);
+			continue;
+		}
+
+		if ((fds[0].revents & POLLNVAL) == POLLNVAL) {
+			LOG_DBG("Socket error: POLLNVAL");
```

Comment:
Why not LOG_ERR? Here and in other places below.

Reply:
because a disconnect is not necessarily an error condition.

Reply:
I'd argue if `POLLNVAL` (someone closed our socket behind our back) or `POLLERR` is not an error condition, but I'll leave it to you, if you prefer LOG_DBG, fine.

Reply:
`POLLHUP` is received on initial pairing when the cloud disconnects the device.
`POLLNVAL` is received when `nrf_cloud_disconnect()` is called.
I'll leave these as debug and let the application print the appropriate message based on the disconnect reason and state.